src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -473,176 +574,223 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
+	    //data is clean now, updating in-memory metadata for an object
+	    e->dirty = false;
+	    //invoke update() with dirty flag set to false, to update in-memory metadata for head
+	    this->update(dpp, new_head_oid_in_cache, 0, 0, e->version, false, y);
+
+	    rgw::d4n::CacheBlock block;
+	    block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
+	    block.cacheObj.objName = c_obj->get_name();
+	    block.size = 0;
+	    block.blockID = 0;
+	    if (c_obj->have_instance()) {
+	      blockDir->get(dpp, &block, null_yield);
+	      if (block.version == c_obj->get_instance()) { //versioned case - update head block entry that has latest version
+		op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
+		if (op_ret < 0) {
+		    ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for head failed!" << dendl;
+		}
+	      }
+	    } else { //non-versioned case
+	      op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
+	      if (op_ret < 0) {
+		  ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for head failed!" << dendl;
+	      }
+	    }
+	    if (c_obj->have_instance()) {
+	      rgw::d4n::CacheBlock instance_block;
+	      instance_block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
+	      instance_block.cacheObj.objName = c_obj->get_oid();
+	      instance_block.size = 0;
+	      instance_block.blockID = 0;
+	      op_ret = blockDir->update_field(dpp, &instance_block, "dirty", "false", null_yield);
+	      if (op_ret < 0) {
+		  ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for instance block failed!" << dendl;
+	      }
+	    }
+	    //remove entry from map and queue, erase_dirty_object locks correctly
+	    erase_dirty_object(dpp, e->key, null_yield);
+	  } else { //end-if std::difftime(time(NULL), e->creationTime) > interval
+	    std::this_thread::sleep_for(std::chrono::milliseconds(interval)); //TODO:: should this time be optimised?
+	  }
         }
-      } else { //non-versioned case
-        op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
-        if (op_ret < 0) {
-            ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for head failed!" << dendl;
-        }
-      }
-      if (c_obj->have_instance()) {
-        rgw::d4n::CacheBlock instance_block;
-        instance_block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
-        instance_block.cacheObj.objName = c_obj->get_oid();
-        instance_block.size = 0;
-        instance_block.blockID = 0;
-        op_ret = blockDir->update_field(dpp, &instance_block, "dirty", "false", null_yield);
-        if (op_ret < 0) {
-            ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for instance block failed!" << dendl;
-        }
+      } else {
+	continue; // e->dirty == false
```

Comment:
why is this continue needed, when will e->dirty be false?

Reply:
Same response as below; should I remove the check entirely?